### ISSUE TITLE
[Proposal/Feature] IntelOwl Plugin Framework

### ISSUE THREAD

**Issue by Eshaan7** [intelowlproject/IntelOwl#249]
I have had this initial idea of having a "IntelOwl plugin framework" which would expose an API (not RESTful/web) that would allow others to create _connectors_/_enrichers_ for IntelOwl **without it necessarily being part of IntelOwl repository/project**. 

So think of it like how we have integrated elasticsearch, it's optional and it automatically syncs job/analysis data with postgres. So a plugin framework would expose some functionalities using _django signals_, for example,
-> a job just got completed and it's result got stored into postgres, so now a event is fired using `django.signals` . This event is then catched by all plugins that are currently enabled and they do some processing of their own, maybe take that analysis data, parse it and send it to another service (_connector_) or replaces the previous raw data with new parsed data in postgres only (_enricher_), or like infinite other ways people want to connect intelowl to some external service

This shouldn't be too hard to roll out on our own or we can use something like [YAPSY](https://yapsy.readthedocs.io/en/latest/index.html#overview) (which allows one to create a plugin based architecture in their python project).

Our goal here is that we would want this plugin framework to be implemented in a way that it works very well for connecting with threat intel platforms.

**Comment by mlodic** [comment]
That would be great! This would be very useful to start to solve some old issues of this project like #12 or #11 and could be also used as a framework to create some kind of post-processing analysis/transformation of the results.

**Comment by Eshaan7** [comment]
An alpha or very basic implementation of this could be:
- create a new django signal `on_job_analysis_finished`
- create a new queue in rabbitmq called `jobs_feed`
- fire this signal when job gets completed
- on the connector side of this new django signal, we submit the `Job` data to the queue.
Flow: new job -> fire signal -> add to queue

Now on the plugin side we (or any plugin developers) can simply use a rabbitmq consumer like kombu to consume this queue.

The advantage of this approach is that the plugins would remain totally decoupled from the application.

**Comment by WasimAkhtarKhan** [comment]
Hi, I am Wasim Akhtar Khan studying Information Technology at BIT Durg, a GSOC aspirant. I want to work on this issue.
I have been using Django since last year and have made some projects.
Looking forward to work with IntelOwl for long time.
It would be really helpful for me if you could suggest some pointers to give me a start.
Thank You
@Eshaan7  @mlodic

**Comment by Eshaan7** [comment]
> Hi, I am Wasim Akhtar Khan studying Information Technology at BIT Durg, a GSOC aspirant. I want to work on this issue.

Welcome to the project, Wasim. This issue would require knowledge of django signals, celery and kombu. 
Please read my above comment about the "alpha/basic implementation", I've explained clearly what needs to be done and how.

**Comment by WasimAkhtarKhan** [comment]
> > Hi, I am Wasim Akhtar Khan studying Information Technology at BIT Durg, a GSOC aspirant. I want to work on this issue.
> 
> Welcome to the project, Wasim. This issue would require knowledge of django signals, celery and kombu.
> Please read my above comment about the "alpha/basic implementation", I've explained clearly what needs to be done and how.

Thank You @Eshaan7

**Comment by Eshaan7** [comment]
Some research:
- [Pluggy](https://pluggy.readthedocs.io/en/latest/) seems like a good alternative to YAPSY. I feel YAPSY has more complex documentation. Last commit on Pluggy was yesterday, whereas last commit on YAPSY was 2 years ago so that's also a thing to consider.
- Very simple "roll-your-own" plugin system: https://www.guidodiepen.nl/2019/02/implementing-a-simple-plugin-framework-in-python/

**Comment by masq** [comment]
Just as a BTW, Python packaging tools has a built-in way of doing a [plugin architecture](https://packaging.python.org/guides/creating-and-discovering-plugins/) via [Entry Points](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points).

The benefit of this model is that in a world where this plugin architecture gets fully implemented, each analyzer could be its own python package that the IntelOwl project can use as analyzers, while they could be completely disjoint from the project, and others could download and use those plugin packages in whatever way they'd like to. A side effect of that is if people want to develop their own plugins for IntelOwl separate from the repo and perhaps private (e.g. enterprise clients with private APIs they want to integrate as an analyzer) they can develop their own plugin in their own repo and have it announce itself as a plugin for IntelOwl to pick up.

Additional benefits would be that it's natively supported by the language/distribution tooling for python packaging. I notice there was concern around projects being abandoned (which is fair concern to have). Unless Python decides to move to 4.x and ditch Entry Points as a feature in that process, I would expect them to stick around for the foreseeable future.

I think one downside to this is that if you "roll your own" plugin architecture, I think it's probably more straightforward and could be easier to implement. Based on my cursory review of the code, it kinda looks like what's in place currently (not that that's bad, it works! but, we are currently discussing other ways to go with it...). I've done a similar smaller scale type thing with Entry Points before, and once I got it working, it was really neat! It was just a lot of scrounging for documentation (I found it wasn't documented _super_ well, and the documentation that was present was somewhat obtuse).

The other downside is just that I _think_ each plugin would _have_ to be its own package. Or they would have to come as a package of several modules (where the one package announces each of the modules/plugins within it). That of course could also be seen as an upside (decoupling, a more "true" plugin architecture).

For a working example of this, I believe the [python documentation](https://packaging.python.org/specifications/entry-points/) points out the [pygments](https://github.com/pygments/pygments) package's plugin architecture is structured this way (making use of Entry Points)

I'm not sure that's the way that you want to go with the plugin architecture here per se, and I'm not sure it's the best path forward for the project, but I figured I'd put it out there in case you hadn't seen it as an option.
